cmd/cli/app/provider/provider_update.go

```diff
@@ -0,0 +1,509 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/stacklok/minder/internal/util/cli"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+var (
+	errMissingProviderName = errors.New("provider name flag is missing")
+	errMissingProject      = errors.New("project flag is missing")
+)
+
+var updateCmd = &cobra.Command{
+	Use:   "update",
+	Short: "Updates a provider's configuration",
+	Long: `The minder provider update command allows a user to update a provider's
+configuration after enrollement.`,
+	RunE: cli.GRPCClientWrapRunE(UpdateProviderCommand),
+}
+
+type parser func(string) (*reflect.Value, error)
+
+func parseBool(val string) (*reflect.Value, error) {
+	v, err := strconv.ParseBool(val)
+	if err != nil {
+		return nil, fmt.Errorf("expected bool, got %s", val)
+	}
+	res := reflect.ValueOf(v)
+	return &res, nil
+}
+
+func parseInt(size int) parser {
+	return func(val string) (*reflect.Value, error) {
+		v, err := strconv.ParseInt(val, 10, size)
+		if err != nil {
+			return nil, fmt.Errorf("expected integer, got %s", val)
+		}
+		res := reflect.ValueOf(v)
+		return &res, nil
+	}
+}
+
+func parseUint(size int) parser {
+	return func(val string) (*reflect.Value, error) {
+		v, err := strconv.ParseUint(val, 10, size)
+		if err != nil {
+			return nil, fmt.Errorf("expected integer, got %s", val)
+		}
+		res := reflect.ValueOf(v)
+		return &res, nil
+	}
+}
+
+func parseFloat(size int) parser {
+	return func(val string) (*reflect.Value, error) {
+		v, err := strconv.ParseFloat(val, size)
+		if err != nil {
+			return nil, fmt.Errorf("expected integer, got %s", val)
+		}
+		res := reflect.ValueOf(v)
+		return &res, nil
+	}
+}
+
+func identity(val string) (*reflect.Value, error) {
+	res := reflect.ValueOf(val)
+	return &res, nil
+}
+
+var (
+	parserMap = map[reflect.Kind]parser{
+		reflect.Bool:    parseBool,
+		reflect.Int32:   parseInt(32),
+		reflect.Int64:   parseInt(64),
+		reflect.Uint32:  parseUint(32),
+		reflect.Uint64:  parseUint(64),
+		reflect.Float32: parseFloat(32),
+		reflect.Float64: parseFloat(64), // double
+		reflect.String:  identity,       // identity
+	}
+)
+
+type configStruct struct {
+	*minderv1.ProviderConfig
+	//nolint:lll
+	GitHub *minderv1.GitHubProviderConfig `json:"github,omitempty" yaml:"github" mapstructure:"github" validate:"required"`
+	//nolint:lll
+	GitHubApp *minderv1.GitHubAppProviderConfig `json:"github_app,omitempty" yaml:"github_app" mapstructure:"github_app" validate:"required"`
```

Comment:
Do you have any thoughts on how would we handle additional custom providers?

Reply:
I haven't looked too much into this, but I believe that custom providers should ship their configuration structure as some form of spec (e.g. JSON Spec) so that every client can perform validation and guarantee type safety when parsing/producing a config object. That would also make it possible to dynamically produce valid attribute names as help strings for example.

Moreover, configuration structure is a problem already, because it's untyped outside of Minder Server, which is why I had to duplicate that struct. I think for our main providers we should be already shipping a spec for the config, and allow (or mandate) custom providers to do the same in the future.